### PR TITLE
DLESyM notebook example

### PR DESCRIPTION
**Pull request by pzharrington**
# Earth2Studio Pull Request

## Description
Add a simple notebook example demonstrating DLESyM usage.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.

### CODE REVIEW

File: earth2studio/models/px/dlesym.py [hunk 1]
```diff
@@ -32,6 +33,7 @@
     OmegaConf = None
     earth2grid = None
 from earth2studio.models.auto import AutoModelMixin, Package
+from earth2studio.models.batch import batch_coords, batch_func
 from earth2studio.models.px.base import PrognosticModel
 from earth2studio.models.px.utils import PrognosticMixin
 from earth2studio.utils import check_extra_imports, handshake_coords, handshake_dim
```
Comment (NickGeneva) [inline earth2studio/models/px/dlesym.py:39]: Does _ATMOS_VARIABLES below need to get updated now that the model computes the derived variables?

File: earth2studio/models/px/dlesym.py [hunk 2]
```diff
@@ -1060,7 +1106,89 @@ def coords_to_ll(self, coords: CoordSystem) -> CoordSystem:
...
+            .roll(lat=len(da_double_interp.lat) - roll_amount_lat, roll_coords=False)
+        )
+
+        return torch.from_numpy(da_triple_interp.values).to(sst.device)
+
+    def _prepare_derived_variables(
+        self, x: torch.Tensor, coords: CoordSystem
+    ) -> tuple[torch.Tensor, CoordSystem]:
+        """
+        Prepare derived variables for the DLESyM model.
```
Comment (NickGeneva) [inline earth2studio/models/px/dlesym.py]: Nit: First line of doc strings should start on same line as the """

E.g. """Prepare derived variables for the DLESyM model.

Reason just being this is what the autodoc extension does by default so all the package follows this

Same with _nan_interpolate_sst

File: earth2studio/models/px/dlesym.py (hunk 1 above)
Comment (pzharrington) [reply earth2studio/models/px/dlesym.py:39]: No, that hardcode is just there for reference -- but this is the core variable set so it remains the same. Only `DLESyMLatLon` accepts the non-derived input vars, and it modifies the input coords accordingly

File: earth2studio/models/px/dlesym.py [hunk 3]
```diff
@@ -1060,7 +1106,89 @@ def coords_to_ll(self, coords: CoordSystem) -> CoordSystem:
...
+        """
+        Custom interpolation to fill NaNs over landmasses in SST data.
+        """
+
+        da_sst = xr.DataArray(sst.cpu().numpy(), dims=coords.keys())
+        da_interp = da_sst.interpolate_na(
+            dim="lon", method="linear", use_coordinate=False
+        )
+
+        # Second pass: roll, interpolate along longitude, and unroll
```
Comment (NickGeneva) [inline earth2studio/models/px/dlesym.py:1120]: I understand why you roll, but would the interpolate in the previous step fill in some in accurate values on the 0 / 359.75 lons? Or does it leave the edges NaNs?

I.e. does the first step on something like [NaN, NaN, 1, 2, NaN, 3, NaN]

Give [NaN, NaN, 1, 2, 2.5, 3, NaN] or [1, 1, 1, 2, 2.5, 3, 3]

File: earth2studio/models/px/dlesym.py [hunk 4]
```diff
@@ -1060,7 +1106,89 @@ def coords_to_ll(self, coords: CoordSystem) -> CoordSystem:
...
+        roll_amount_lon = int(len(da_interp.lon) / 2)
+        da_double_interp = (
+            da_interp.roll(lon=roll_amount_lon, roll_coords=False)
+            .interpolate_na(dim="lon", method="linear", use_coordinate=False)
+            .roll(lon=len(da_interp.lon) - roll_amount_lon, roll_coords=False)
+        )
+
+        # Third pass do a similar roll along latitude
```
Comment (NickGeneva) [inline earth2studio/models/px/dlesym.py:1128]: Wait roll on lat? Huh? Shouldnt it be reflective padding? Roll would put south poll against the north pole?

Comment (NickGeneva) [reply earth2studio/models/px/dlesym.py:1128]: Wait is this healpix here?

Comment (pzharrington) [reply earth2studio/models/px/dlesym.py:1128]: I pulled this method directly from the DLESyM project data prep code so haven't thought much about the behavior. It's certainly weird, but this is what the model is trained with. Initializing the forecast with a simple constant-fill over landmasses causes the forecast to eventually diverge

Comment (pzharrington) [reply earth2studio/models/px/dlesym.py:1128]: The polar behavior seems to work out because both poles are...cold. Goes from this:
image
to this:
image

File: earth2studio/models/px/dlesym.py [hunk 5]
```diff
@@ -972,9 +1011,16 @@ def input_coords(self) -> CoordSystem:
         """
         coords = super().input_coords()
         coords = self.coords_to_ll(coords)
+
+        # Modify to use the base variables instead of the derived variables
+        input_variables = [
```
Comment (NickGeneva) [inline earth2studio/models/px/dlesym.py:1017]: dis-regard my question above about the variables getting updated

File: earth2studio/models/px/dlesym.py (hunk 4 above)
Comment (NickGeneva) [reply earth2studio/models/px/dlesym.py:1128]: I guess north an south pole are cold, but if this is from the code base good enough for me

**Comment by NickGeneva** [comment]
LGTM, going to pull and try running the example myself rq

File: examples/14_dlesym_example.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# %%
+"""
+Running DLESyM Inference
+===============================
```
Comment (NickGeneva) [inline examples/14_dlesym_example.py]: Sphinx nit: "===============================" same length as "Running DLESyM Inference"

File: examples/14_dlesym_example.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+    "Lat-lon output variable names: ", model_ll.output_coords(in_coords_ll)["variable"]
+)
+print("HEALPix input variable names: ", in_coords_hpx["variable"])
+print(
+    "HEALPix output variable names: ",
+    model_hpx.output_coords(in_coords_hpx)["variable"],
+)
+# %%
+# Making predictions, regridding, and selecting outputs
+# ---------------
```
Comment (NickGeneva) [inline examples/14_dlesym_example.py]: Sphinx Nit: "---------------" should be same length as title

Making predictions, regridding, and selecting outputs > Capitalize words like a title

File: examples/14_dlesym_example.py [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+- How to use the model API to generate a forecast
+- How to use the output selection and regridding methods to select appropriate data
+- How to use the `DLESyMLatLon` model with earth2studio workflows
+"""
+
+# %%
+# Set Up
+# ------
+# The first step is fetching appropriate input data for the model. The ERA5 data sources
+# in earth2studio provide data on the lat/lon grid, so have two options:
```
Comment (NickGeneva) [inline examples/14_dlesym_example.py:45]: "so there are two options" or something

File: earth2studio/models/px/dlesym.py [hunk 9]
```diff
@@ -32,6 +33,7 @@
     OmegaConf = None
```
Comment (NickGeneva) [inline earth2studio/models/px/dlesym.py:33]: I think you need to add `hydra-core` to the dependencies for dlesym, modulus doesnt install this by default but seems an arch in the model uses the instantiate method

Comment (NickGeneva) [reply earth2studio/models/px/dlesym.py:33]: uv add hydra-core --optional dlesym

File: examples/14_dlesym_example.py [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+#  -  use the :py:class:`earth2studio.models.DLESyMLatLon` model. This version of DLESyM
```
Comment (NickGeneva) [inline examples/14_dlesym_example.py]: RST error, extra space in front of the "-", should be

```
# The first step is fetching appropriate input data for the model. The ERA5 data sources
# in earth2studio provide data on the lat/lon grid, so have two options:
#
# -  use the :py:class:`earth2studio.models.DLESyMLatLon` model. This version of DLESyM
#     accepts inputs on the lat/lon grid and regrids them to the HEALPix grid internally,
#     before returning the output regridded back to the lat/lon grid. This is the
#     recommended approach for most users as it can be used directly with earth2studio
#     data sources and workflows, since it performs regridding and pre-processing
#     internally.
# -  use the :py:class:`earth2studio.models.DLESyM` model, and handle the regridding of
#     input lat/lon data ourselves. Since the model uses some derived variables which
#     are not provided by the data source, we would also need to prepare these derived
#     variables ourselves.
```

File: earth2studio/models/px/dlesym.py (hunk 9 above)
Comment (pzharrington) [reply earth2studio/models/px/dlesym.py:33]: Interesting, yeah that's probably buried in the `Module.from_checkpoint` instantiation and the dep was covered by using the physicsnemo container. Will add

File: examples/14_dlesym_example.py [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+# ---------------
+# Let's now pull some example data and make predictions with the model. As the
+# data source provides lat/lon data, we can use the :py:class:`earth2studio.models.DLESyMLatLon` model.
+
+# In addition, we demonstrate how to use the regridding utilities provided by
+# `DLESyMLatLon` to regrid onto the HEALPix grid. The :py:class:`earth2studio.models.DLESyM`
+# model can then be used directly with the HEALPix data.
+
+# Finally, a key aspect of the DLESyM model is that it makes predictions for the atmosphere
+# and ocean components at different timesteps, becasue the atmosphere is faster-evolving
```
Comment (NickGeneva) [inline examples/14_dlesym_example.py]: becasue -> because

File: examples/14_dlesym_example.py [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+    model_hpx.output_coords(in_coords_hpx)["variable"],
+)
+# %%
+# Making predictions, regridding, and selecting outputs
+# ---------------
+# Let's now pull some example data and make predictions with the model. As the
+# data source provides lat/lon data, we can use the :py:class:`earth2studio.models.DLESyMLatLon` model.
+
+# In addition, we demonstrate how to use the regridding utilities provided by
+# `DLESyMLatLon` to regrid onto the HEALPix grid. The :py:class:`earth2studio.models.DLESyM`
```
Comment (NickGeneva) [inline examples/14_dlesym_example.py]: earth2studio.models.px.DLESyM

File: examples/14_dlesym_example.py [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+print("HEALPix input variable names: ", in_coords_hpx["variable"])
+print(
+    "HEALPix output variable names: ",
+    model_hpx.output_coords(in_coords_hpx)["variable"],
+)
+# %%
+# Making predictions, regridding, and selecting outputs
+# ---------------
+# Let's now pull some example data and make predictions with the model. As the
+# data source provides lat/lon data, we can use the :py:class:`earth2studio.models.DLESyMLatLon` model.
```
Comment (NickGeneva) [inline examples/14_dlesym_example.py]: earth2studio.models.px.DLESyMLatLon

File: examples/14_dlesym_example.py [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+# ------
+# The first step is fetching appropriate input data for the model. The ERA5 data sources
+# in earth2studio provide data on the lat/lon grid, so have two options:
+#  -  use the :py:class:`earth2studio.models.DLESyMLatLon` model. This version of DLESyM
+#     accepts inputs on the lat/lon grid and regrids them to the HEALPix grid internally,
+#     before returning the output regridded back to the lat/lon grid. This is the
+#     recommended approach for most users as it can be used directly with earth2studio
+#     data sources and workflows, since it performs regridding and pre-processing
+#     internally.
+#  -  use the :py:class:`earth2studio.models.DLESyM` model, and handle the regridding of
```
Comment (NickGeneva) [inline examples/14_dlesym_example.py]: earth2studio.models.px.DLESyM

File: examples/14_dlesym_example.py (hunk 10 above)
Comment (NickGeneva) [reply examples/14_dlesym_example.py]: earth2studio.models.px.DLESyMLatLon

File: examples/14_dlesym_example.py [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+print(
+    "HEALPix output variable names: ",
+    model_hpx.output_coords(in_coords_hpx)["variable"],
+)
+# %%
+# Making predictions, regridding, and selecting outputs
+# ---------------
+# Let's now pull some example data and make predictions with the model. As the
+# data source provides lat/lon data, we can use the :py:class:`earth2studio.models.DLESyMLatLon` model.
+
```
Comment (NickGeneva) [inline examples/14_dlesym_example.py]: These empty lines here need a "#"

I.e.

```
# Making Predictions, Regridding, and Selecting Outputs
# -----------------------------------------------------
# Let's now pull some example data and make predictions with the model. As the
# data source provides lat/lon data, we can use the :py:class:`earth2studio.models.px.DLESyMLatLon`
# model.
#
# In addition, we demonstrate how to use the regridding utilities provided by
# `DLESyMLatLon` to regrid onto the HEALPix grid. The :py:class:`earth2studio.models.DLESyM`
# model can then be used directly with the HEALPix data.
#
# Finally, a key aspect of the DLESyM model is that it makes predictions for the atmosphere
# and ocean components at different timesteps, becasue the atmosphere is faster-evolving
# than the ocean. The atmosphere is predicted every 6 hours, while the ocean is only
# predicted every 48 hours. Thus, not all output lead times are valid for the ocean
# component. For convenience, we can use a method that selects only the valid outputs
# for each of the atmosphere and ocean components.
```

File: examples/14_dlesym_example.py [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+)
+print(
+    "Ocean outputs (variables, lead_time [hrs]):",
+    y_ocean_coords["variable"],
+    y_ocean_coords["lead_time"].astype("timedelta64[h]"),
+)
+
+# %%
+# Model iteration for longer forecasts
+# ------------------------------
```
Comment (NickGeneva) [inline examples/14_dlesym_example.py]: Nit: fix under score and title caps

File: examples/14_dlesym_example.py [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+    if i > 0:  # Don't retrieve the first step as it is the initial condition
+        x_atmos, x_atmos_coords = model_ll.retrieve_valid_atmos_outputs(x, x_coords)
+        x_ocean, x_ocean_coords = model_ll.retrieve_valid_ocean_outputs(x, x_coords)
+
+print(f"Completed forecast with {n_steps} steps")
+
+
+# %%
+# Using built-in deterministic workflow
+# ------------------------------------
```
Comment (NickGeneva) [inline examples/14_dlesym_example.py]: Nit

File: examples/14_dlesym_example.py [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+grid for the physical variables of interest. The model also uses some derived input
+variables which are not provided by standard data sources but can be computed
+from the standard variables.
+
+In this example you will learn:
+
+- How to instantiate the DLESyM model
+- How to use the model API to generate a forecast
+- How to use the output selection and regridding methods to select appropriate data
+- How to use the `DLESyMLatLon` model with earth2studio workflows
```
Comment (NickGeneva) [inline examples/14_dlesym_example.py]: Just DLESyMLatLon

\` makes it italic for some reason lol

File: examples/14_dlesym_example.py (hunk 10 above)
Comment (NickGeneva) [reply examples/14_dlesym_example.py]: Update with improved spacing:

```
# %%
# Set Up
# ------
# The first step is fetching appropriate input data for the model. The ERA5 data sources
# in earth2studio provide data on the lat/lon grid, so have two options:
#
# - Use the :py:class:`earth2studio.models.px.DLESyMLatLon` model. This version of DLESyM
#   accepts inputs on the lat/lon grid and regrids them to the HEALPix grid internally,
#   before returning the output regridded back to the lat/lon grid. This is the
#   recommended approach for most users as it can be used directly with earth2studio
#   data sources and workflows, since it performs regridding and pre-processing
#   internally.
# - Use the :py:class:`earth2studio.models.px.DLESyM` model, and handle the regridding of
#   input lat/lon data ourselves. Since the model uses some derived variables which
#   are not provided by the data source, we would also need to prepare these derived
#   variables ourselves.
#
# Let's load both of these models and inspect the expected input coordinates for each.
# Also note the input and output variable set for each model.

# %%
```

**Comment by NickGeneva** [comment]
Example looks fantastic! added some minor changes needed.

Use `make docs` to test yourself to see if the rst looks good. This is great

**Review by NickGeneva — APPROVED** [review]
Approve to unblock

File: earth2studio/models/px/dlesym.py [hunk 19]
```diff
@@ -85,10 +87,11 @@ class DLESyM(torch.nn.Module, AutoModelMixin, PrognosticMixin):
...
-            atmos_outputs, atmos_coords = model.retrieve_valid_atmos_outputs(x, coords)
-            ocean_outputs, ocean_coords = model.retrieve_valid_ocean_outputs(x, coords)
-            ...
+            if step > 0:
+                # Valid atmos and ocean predictions with their respective coordinates extracted below
+                atmos_outputs, atmos_coords = model.retrieve_valid_atmos_outputs(x, coords)
+                ocean_outputs, ocean_coords = model.retrieve_valid_ocean_outputs(x, coords)
+                ...
 
     Note
```
Comment (NickGeneva) [inline earth2studio/models/px/dlesym.py:96]: Last Nit:

Update to (align the spaces with the bullets):

```
    Note
    ----
    For more information about this model see:

    - https://arxiv.org/abs/2409.16247
    - https://arxiv.org/abs/2311.06253v2

    For more information about the HEALPix grid see:
    
    - https://github.com/NVlabs/earth2grid
    
```
